core/src/consensus/tendermint/worker.rs

```diff
@@ -817,47 +878,49 @@ impl Worker {
             self.last_two_thirds_majority =
                 TwoThirdsMajority::from_message(message.on.step.view, message.on.block_hash);
         }
+
+        if vote_step.step == Step::Precommit
+            && self.height == vote_step.height
+            && message.on.block_hash.is_some()
+            && has_enough_aligned_votes
+        {
+            if self.can_move_from_commit_to_propose() {
+                let height = self.height;
+                self.move_to_height(height + 1);
+                self.move_to_step(TendermintState::Propose, is_restoring);
+                return
+            }
+
+            let block_hash = message.on.block_hash.expect("Upper if already checked block hash");
+            if !self.step.is_commit() {
+                self.move_to_step(
+                    TendermintState::Commit {
+                        block_hash,
+                        view: vote_step.view,
+                    },
+                    is_restoring,
+                );
+                return
+            }
+        }
+
         // Check if it can affect the step transition.
         if self.is_step(message) {
             let next_step = match self.step {
                 TendermintState::Precommit if message.on.block_hash.is_none() && has_enough_aligned_votes => {
                     self.increment_view(1);
-                    Some(Step::Propose)
-                }
-                TendermintState::Precommit if has_enough_aligned_votes => {
-                    let bh = message.on.block_hash.expect("previous guard ensures is_some; qed");
-                    if self.client().block(&BlockId::Hash(bh)).is_some() {
-                        // Commit the block, and update the last confirmed view
-                        self.save_last_confirmed_view(message.on.step.view);
-
-                        // Update the best block hash as the hash of the committed block
-                        self.client().update_best_as_committed(bh);
-                        Some(Step::Commit)
-                    } else {
-                        cwarn!(ENGINE, "Cannot find a proposal which committed");
-                        self.increment_view(1);
-                        Some(Step::Propose)
-                    }
```

Comment:
Before this PR, if the committed block is not imported when moving to the Commit state, the node moves to the next round.
After this PR, the node moves to Commit state and waits for the block.